### PR TITLE
Fix issue 6652: Raise `aiohttp.ServerFingerprintMismatch` exception o…

### LINKED ISSUE #6652: Server certificate fingerprint is not verified when request through http proxy
### Describe the bug

When making request as below to try to verify server certificate fingerprint 'digest' during http request:

```
tcp_connector = TCPConnector(ssl=aiohttp.Fingerprint(digest))
async with aiohttp.ClientSession(headers=headers, connector=tcp_connector, trust_env=True) as session:
    async with session.request(...) as r:
        ....
```

If the request go directly to the server, the server fingerprint is verified successfully: when the 'digest' matches the server certificate fingerprint digest, the request succeed, when it doesn't match, a 'aiohttp.ServerFingerprintMismatch' exception is thrown.
While if the request need to go through a http proxy (with env: HTTP_PROXY, HTTPS_PROXY set as the proxy), the server fingerprint is not verified: whether the 'digest' matches the server certificate fingerprint digest or not, the request will always succeed, no 'aiohttp.ServerFingerprintMismatch' exception is thrown.

### To Reproduce

1. setup a "http server" and a "http proxy" through which the "http server" can be requested.
2. write a simple program as above:
```
tcp_connector = TCPConnector(ssl=aiohttp.Fingerprint(digest))
async with aiohttp.ClientSession(headers=headers, connector=tcp_connector) as session:
    async with session.request(...) as r:
        ....
```
set the 'digest' as a mismatch one as the http server's certificate fingerprint digest
3. the request can be made successfully

### Expected behavior

The request should fail with a 'aiohttp.ServerFingerprintMismatch' exception.

### Logs/tracebacks

```python-traceback
NONE
```

### Python Version

```console
$ python3 -V
Python 3.8.10
```

### aiohttp Version

```console
$ python3 -m pip show aiohttp
Name: aiohttp
Version: 3.8.1
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author:
Author-email:
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires: aiosignal, multidict, frozenlist, attrs, charset-normalizer, async-timeout, yarl
Required-by: uniconlib, aiohttp-jinja2
```

### multidict Version

```console
$ python3 -m pip show multidict
Name: multidict
Version: 6.0.2
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires:
Required-by: yarl, aiohttp
```

### yarl Version

```console
$ python3 -m pip show yarl
Name: yarl
Version: 1.7.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.8/dist-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

$ cat /etc/issue
Ubuntu 20.04.4 LTS \n \l

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by gangj**
Fix issue 6652: Raise `aiohttp.ServerFingerprintMismatch` exception on client-side if request through http proxy with mismatching server fingerprint digest

## What do these changes do?
Fix issue 6652: Raise `aiohttp.ServerFingerprintMismatch` exception on client-side if request through http proxy with mismatching server fingerprint digest

## Are there changes in behavior for the user?
Users will be able to verify server certificate fingerprint for requests which go through http proxy

## Related issue number
Fixes #6652

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by Dreamsorcerer** [comment]
Can we add a test to cover this case?

**Comment by gangj** [comment]
> Can we add a test to cover this case?

OK, will have a try.

**Comment by gangj** [comment]
> Can we add a test to cover this case?

I had added a test: "test_https_connect_fingerprint_mismatch", please help to have a review, thanks!

**Comment by gangj** [comment]
Hi, 

There remains one item: "Documentation reflects the changes" in the checklist not done yet. But I checked related document, seems there is no need to add document for this bug fix, any ideas?

Thanks!

File: aiohttp/connector.py [hunk 1]
```diff
@@ -1044,6 +1044,14 @@ async def _start_tls_connection(
                     # chance to do this:
                     underlying_transport.close()
                     raise
+                if isinstance(tls_transport, asyncio.Transport):
+                    fingerprint = self._get_fingerprint(req)
+                    if fingerprint:
+                        try:
+                            fingerprint.check(tls_transport)
+                        except ServerFingerprintMismatch:
+                            tls_transport.close()
```
Comment (bdraco) [inline aiohttp/connector.py:1142]: Do we need to consider `_cleanup_closed_transports` here?

File: aiohttp/connector.py [hunk 2]
```diff
@@ -1133,6 +1133,16 @@ async def _start_tls_connection(
...
+                            if not self._cleanup_closed_disabled:
+                                self._cleanup_closed_transports.append(transp)
```
Comment (Dreamsorcerer) [inline aiohttp/connector.py]:
```suggestion
                                self._cleanup_closed_transports.append(tls_transport)
```

File: tests/test_proxy.py [hunk 3]
```diff
@@ -415,6 +416,99 @@ async def make_conn() -> aiohttp.TCPConnector:
                             self.loop.run_until_complete(req.close())
                 self.loop.run_until_complete(connector.close())
 
+    @pytest.parametrize("cleanup", (True, False))
```
Comment (Dreamsorcerer) [inline tests/test_proxy.py]:
```suggestion
    @pytest.mark.parametrize("cleanup", (True, False))
```

File: tests/test_proxy.py [hunk 4]
```diff
@@ -5,10 +5,11 @@
 import unittest
 from unittest import mock
 
+import pytest
```
Comment (Dreamsorcerer) [inline tests/test_proxy.py]:
```suggestion
```

File: tests/test_proxy.py [hunk 5]
```diff
@@ -415,6 +415,106 @@ async def make_conn() -> aiohttp.TCPConnector:
...
                 self.loop.run_until_complete(connector.close())
 
+    @mock.patch("aiohttp.connector.ClientRequest")
+    @mock.patch(
+        "aiohttp.connector.aiohappyeyeballs.start_connection",
+        autospec=True,
+        spec_set=True,
+    )
+    def test_https_connect_fingerprint_mismatch(
+        self, start_connection: mock.Mock, ClientRequestMock: mock.Mock, cleanup: bool
```
Comment (Dreamsorcerer) [inline tests/test_proxy.py]:
```suggestion
        self, start_connection: mock.Mock, ClientRequestMock: mock.Mock
```

**Pull request merged**